ext/nnstreamer/tensor_filter/tensor_filter_python_api.c

```diff
@@ -0,0 +1,217 @@
+TensorShape_getType(TensorShapeObject *self, PyObject *args) {
+  return Py_BuildValue("O", self->type);
+}
+
+/** 
+ * @brief new callback for custom type object 
+ * @param self : Python type object 
+ * @param args : arguments for the method
+ * @param kw : keywords for the arguments
+ */
+static PyObject *
+TensorShape_new(PyTypeObject *type, PyObject *args, PyObject *kw) {
+  TensorShapeObject *self = (TensorShapeObject *) type->tp_alloc(type, 0);
+
+  g_assert(self);
+
+  // Assign default values
+  self->dims = PyList_New(0);
+  self->type = PyArray_DescrFromType(NPY_UINT8);
+  Py_XINCREF(self->type);
+
+  return (PyObject *) self;
+}
+
+/** 
+ * @brief init callback for custom type object 
+ * @param self : Python type object 
+ * @param args : arguments for the method
+ * @param kw : keywords for the arguments
+ */
+static int
+TensorShape_init(TensorShapeObject *self, PyObject *args, PyObject *kw) {
+  char *keywords[] = {"dims", "type", NULL};
+  PyObject *dims = NULL;
+  PyObject *type = NULL;
+
+  if (!PyArg_ParseTupleAndKeywords(args, kw, "|OO", keywords, &dims, &type))
+    return -1;
+
+  if (dims) {
+    PyObject *none = PyObject_CallMethod((PyObject*) self, (char*) "setDims", "O", dims);
+    Py_XDECREF(none);
+  }
+
+  if (type) {
+    PyArray_Descr* dtype;
+    if (PyArray_DescrConverter(type, &dtype) != NPY_FAIL) {
+      // swap 'self->type' 
+      Py_XDECREF(self->type);
+      self->type = dtype;
+      Py_XINCREF(dtype);
+    } else
+      g_critical("Wrong data type");
+  }
+
+  return 0;
+}
+
+/** 
+ * @brief dealloc callback for custom type object 
+ * @param self : Python type object 
+ */
+static void
+TensorShape_dealloc(TensorShapeObject *self) {
+  Py_XDECREF(self->dims);
+  Py_XDECREF(self->type);
+  Py_TYPE(self)->tp_free((PyObject *) self);
+}
+
+/** @brief members for custom type object */
+static PyMemberDef TensorShape_members[] = {
+  {"dims", T_OBJECT_EX, offsetof(TensorShapeObject, dims), 0, NULL},
+  {"type", T_OBJECT_EX, offsetof(TensorShapeObject, type), 0, NULL}
+};
+
+/** @brief methods for custom type object */
+static PyMethodDef TensorShape_methods[] = {
+  {"setDims", (PyCFunction)TensorShape_setDims, METH_VARARGS | METH_KEYWORDS, NULL},
+  {"getDims", (PyCFunction)TensorShape_getDims, METH_VARARGS | METH_KEYWORDS, NULL},
+  {"getType", (PyCFunction)TensorShape_getType, METH_VARARGS | METH_KEYWORDS, NULL},
+  {NULL, NULL, 0, NULL}
+};
+
+/** @brief Structure for custom type object */
+static PyTypeObject TensorShapeType = {
+  PyVarObject_HEAD_INIT(NULL, 0)
+  .tp_name = "nnstreamer_python.TensorShape",
+  .tp_doc = "TensorShape type",
+  .tp_basicsize = sizeof(TensorShapeObject),
+  .tp_itemsize = 0,
+  .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE,
+  .tp_new = TensorShape_new,
+  .tp_init = (initproc) TensorShape_init,
+  .tp_dealloc = (destructor) TensorShape_dealloc,
+  .tp_members = TensorShape_members,
+  .tp_methods = TensorShape_methods
+};
+
+#if PY_VERSION_HEX >= 0x03000000
```

Comment:
Be careful. ```PY_VERSION_HEX ``` is determined at compile time.

Reply:
This source file will be compiled twice. 
(Refer to /ext/nnstreamer/tensor_filter/meson.build)

Reply:
I've checked both libraries are generated and work well.
In fact, this file is compiled twice (according to meson.build).

Reply:
If it is used twice, each for python 2 and 3, it is fine.